### ISSUE TITLE
Make editor more compact and detailed

### ISSUE THREAD

**Issue by AndreaCatania** [godotengine/godot-proposals#9]
Looking at the blender editor you can see a lot of information that are ready to be used without any scroll. Each section like the inspector or the scene navigation is really detailed since no space is wasted.

Opposed to this the Godot editor uses too much spaces to represent even a checkbox, and this make the editor to lock at first glance too spartan and later too uncomfortable. And is easy find yourself scrolling a lot during editor usage.

**(Down a comparison)**

The fact that Godot editor looks so spartan is due to two main factors in my opinion.
- The first one is due to the "General HUD generation" that allow to generate the HUD of any node fitting always in a good way all parameters no matter the type and the real space required.
- The other problem instead is the font type and font size.

Considering these two factors I think that the first one is good to have and difficult to improve, but the second one is easy to improve and changing is a benefit.

For this reason I want to propose to use a better font like the one used by blender (that is called blender and probably is OS) and then match the same font size used in blender.

Thanks to this change we may improve the look and feel of the editor a lot.

-----
# Some measurements:
![Screenshot2](https://user-images.githubusercontent.com/8342599/57845994-85511880-77d3-11e9-8f55-19f2bce1fee3.png)

**Comment by girng** [comment]
I don't know, I always hated the blender inspector because everything was so scrunched together. However, that's not saying this isn't a good idea. Different folks for different folks I guess

**Comment by samdze** [comment]
I agree, the Godot UI is too distended to the point it feels relatively confusing and hard to navigate.

I always notice that when I pass from Godot to Unity or vice versa:

- In Unity, a lot of data is displayed in relatively little space, in a way that is still readable and makes me feel under complete control of the situation, I can easily see what I'm interacting with, the context in which I'm doing it, a lot of other data that might interest me etc... (also because de editor panels are rearranged at will)

- In Godot, UI elements are big but it's not just a matter of size, the way they are displayed and organized makes it harder to track down what you're interacting with or what you'd like to interact with, the least amount of data accessible at a glance makes things worse.

**Comment by Calinou** [comment]
For the record, this is how the editor looks like at font size 13 with a window size of 1920×1080 (including window borders, which I displayed for completeness):

![image](https://user-images.githubusercontent.com/180032/64364124-8ae89d80-d012-11e9-87b5-2669cefd1436.png)

This is with font size 14 (the current default value):

![image](https://user-images.githubusercontent.com/180032/64364164-9e940400-d012-11e9-9662-5c294c206dcf.png)

Decreasing the font size increases the vertical screen real estate significantly, but care needs to be taken as to not make list elements too hard to click. For example, when the editor font size is set to 13, menu buttons become difficult to target:

![image](https://user-images.githubusercontent.com/180032/64364280-cedba280-d012-11e9-8271-aa98d66021d7.png)

This could be solved by increasing the vertical separation at lower font sizes.

**Comment by aaronfranke** [comment]
Considering Blender recently redesigned their UI with Blender 2.8, if you're comparing with Blender, this proposal would be incomplete without a comparison with Blender 2.8's UI.

Overall I quite like Godot's current UI, but there are certainly ways to improve it. Checkboxes are an obvious example, currently each takes up the entire inspector width.

**Comment by Calinou** [comment]
I don't know how checkbox design could be improved as to take less space in the inspector. Cramming multiple checkboxes on the same row sounds difficult to do, unless you resize the inspector to be very wide. And even then, you may end up displaying completely unrelated properties on the same row.

**Comment by samdze** [comment]
I think the UI issue is not just a matter of sizes.
How Godot arranges visual elements creates this subtle sensation of disorder that I don't feel using Unity, for example.
The text and other elements' size contributes to this feeling, but it's just one of the actors in my opinion.

As food for thought:

1. The inspector generally gives me the idea that it is very limited in functionality (even if it is not)
maybe due to its limited variety when displaying different types of controls, they almost look all the
same, I think this is one of the reasons why I don't feel like I'm in full control.
2. I find it a little confusing as well:
property names and categories aren't visually organized and represented in the most readable and distinct way. (and I'd like if they could use more horizontal space effectively)

Maybe this is material for another proposal, but just wanted to explain why changing sizes is not enough imo.

**Comment by Calinou** [comment]
One way to improve the situation would be to align property names to the right, while keeping the property values left-aligned. This is commonly done in column-based interfaces to make two columns easier to read (look at music players).

**Comment by Jummit** [comment]
related: godotengine/godot#2207.

**Comment by samdze** [comment]
@Jummit the concept proposed there is very nice and also much more well organized and displayed than the current theme.
I hope Godot will go more towards that direction in the future, with 3.0 progress has been made, I'm sure there are ways to get even closer to that concept without sacrificing usability.

@Calinou I'm not sure it would be enough, I think there are other graphic elements that should be moved, resized and/or changed in shape/color. The concept godotengine/godot#2207 really gets this right.

**Comment by samdze** [comment]
@girng Yeah, interesting. Why do you think it gives you more confidence?
To me the very similar look makes them appear as similar in functionality too, and the typical inspector control gives very little interactivity.

**Comment by girng** [comment]
> @girng Yeah, interesting. Why do you think it gives you more confidence?
> To me the very similar look makes them appear as similar in functionality too, and the typical inspector control gives very little interactivity.

Lol, not really sure tbh, it just does because I know the inspector will be uniform dependent on any possible future node I click. This eases my mind I guess, hard to explain.

**Comment by AndreaCatania** [comment]
Check this:
![Screenshot from 2019-10-07 16-43-46](https://user-images.githubusercontent.com/8342599/66321947-a4d31400-e921-11e9-8e6c-6bebe92351a4.png)

I've only changed the font to `Source Code Pro` and set the size to 12.

With only that change the inspector is much more compact.

**Comment by dreamsComeTrue** [comment]
> With only that change the inspector is much more compact.

...but, despite the fact that font looks just ugly there, each line looks like it would clamp with the bottom neighbour :( Maybe selection of right font would do the trick for you?

**Comment by pycbouh** [comment]
Well, a proper font can make or break UI, this is true. I would like to note that we can achieve more concise look with the same font size using a shorter font.

I was trying out different Google Fonts to find a better fit, and [Work Sans](https://fonts.google.com/specimen/Work+Sans) seems to be pretty and clear even at smaller sizes. I went for a thicker one (Medium), but it looks fine on regular thickness as well. Thicker style may work better, if changing font size from 14 to 13 or 12. Here is an overall comparison:
[Editor Settings, default](https://user-images.githubusercontent.com/11782833/70329857-612b5680-184d-11ea-8291-dad7ac80afe1.png)
[Editor Settings, WorkSans Regular](https://user-images.githubusercontent.com/11782833/70329874-6b4d5500-184d-11ea-8aa4-fa65ada43330.png)
[Editor Settings, WorkSans Medium](https://user-images.githubusercontent.com/11782833/70329871-6b4d5500-184d-11ea-9bc4-010d95e3c31b.png)

As you can see, it's shorter, but wider. Which is a compromise: shorter font gives more vertical space, while eating more horizontal space to preserve clarity. Personally, I'm okay with that. In screenshots above this font gives a couple more lines of content on screen. Here's the Inspector with similar results:
[Inspector, default](https://user-images.githubusercontent.com/11782833/70330000-b0718700-184d-11ea-8675-a02d1b25101e.png)
[Inspector, WorkSans Regular](https://user-images.githubusercontent.com/11782833/70330002-b0718700-184d-11ea-87d5-98d0b3a03560.png)
[Inspector, WorkSans Medium](https://user-images.githubusercontent.com/11782833/70330001-b0718700-184d-11ea-9fa0-35abb1cd2307.png)

I don't feel like it adds more clutter with reduced vertical spacing, at least _between_ blocks of text. _Inside_ blocks, however, there is still a problem, that smaller font makes some buttons harder to click. I suggest, that in some areas font should be adjusted independently from the main font size and style. There might not be a universal solution, that is as good for Inspector as it is for the menus. Here's the main menu (the difference is 2 pixels):
[Main Menu, default](https://user-images.githubusercontent.com/11782833/70330343-71900100-184e-11ea-93af-d9a8f2f912dd.png)
[Main Menu, WorkSans Regular](https://user-images.githubusercontent.com/11782833/70330348-72289780-184e-11ea-9623-2a82142897ac.png)
[Main Menu, WorkSans Medium](https://user-images.githubusercontent.com/11782833/70330346-72289780-184e-11ea-936e-900f32e4b266.png)

I would like to add, that for the most part, Godot Editor is fine on smaller screens. I mainly use it on a 15.6" laptop, and while there is a criminally small amount of space for script editor because of the vertical list of script files (tabs would be much better, like in every other code editor), I wouldn't say, that the work is crippled or completely uncomfortable.

**Comment by Calinou** [comment]
@pycbouh Character set support is an important criterion for us, as we'd like the editor to support as many languages as possible (without resorting to too many fallback fonts).

**Comment by pycbouh** [comment]
@Calinou Yes, obviously, we need a font with full character set. I'm not advertising this particular font, this is just what I have found as an interesting alternative among the first 20-30 fonts available from Google and haven't checked if it did support characters beyond ASCII.

As to the stretched look, I think it's a matter of familiarity. Current one looks too stretched vertically, given a comparison. It's hard to have a font be smaller vertically and not use extra space horizontally, if we want it to remain readable.

**Comment by pycbouh** [comment]
@Calinou Here's another example. [Exo 2](https://fonts.google.com/specimen/Exo+2) looks like a font for a sci-fi game, which may be a turn-off, but even at 15 it allows for more information to be displayed on screen. At 14 it turns less readable IMO, but the amount of space available is enormous compared to Noto Sans.

[Editor Settings, Exo 2 at size 14](https://user-images.githubusercontent.com/11782833/70337894-f6821700-185c-11ea-9e8f-7523edf39a61.png)
[Editor Settings, Exo 2 at size 15](https://user-images.githubusercontent.com/11782833/70337386-04836800-185c-11ea-8b46-e6b43460cdce.png)

We can all look for a more proper font, but to do so we need to establish, first and foremost, if we indeed see the benefit of changing it. If the maintainer team does not think that the change is needed or is beneficial, at least in principle, then this research is moot. And then we need to formalize requirements for such font, like character set support, font style, preferable dimensions, etc.

**Comment by Jummit** [comment]
I think the main reason why the concept from godotengine/godot#2207 looks visually pleasing is that it gets rid of the boxy outlines everything in Godot has. It also uses round edges for some elements and gets rid of some clutter, like the lines in the scene tree.

**Comment by Calinou** [comment]
@Jummit That's precisely the style I've been using for my Godot editor theme redesign :slightly_smiling_face:

**Comment by samdze** [comment]
@Jummit Yes, all the current clutter, the unnecessary outlines and borders, the blocky style, make for a unpleasant UI compared to that one. The font then really adds to that, very elegant.

I think Godot should try to hide as much as possible the underlying structure that is obviously composed by block shapes, and this can be achieved, as already said, removing lines, rounding a few key UI shapes, highlighting UI elements in a smart and pleasant way and choosing a new coherent font.
Finally, a work of spaces and margins rebalancing could be needed as well.

@Calinou Can we see you progress?

**Comment by pycbouh** [comment]
By the way, since we discussed fonts in December I've adapted IBM Plex Sans as my main GUI font for the editor. It does not have a dramatic effect on the editor space, but at medium thickness it looks much easier on the eye.

[Default Noto Sans](https://user-images.githubusercontent.com/11782833/76411350-b1f9a080-63a2-11ea-9e18-4bc9ab7eab80.png)
[IBM Plex Sans (Medium)](https://user-images.githubusercontent.com/11782833/76411377-bf168f80-63a2-11ea-802b-997c7990a0ef.png)

It's distributed under SIL OFL and has less glyphs than Noto, but European languages, including Cyrillic alphabet, should be covered.

-----
Now, looking at Noto Sans' glyphs in Google Fonts does not feel as bad as it does in the editor. I can think of two things: there is some issue with rendering scaled down fonts in Godot; or Medium does it for me, and Noto does not have a Medium style.

While thickness may be subjective and down to one's (bad) eyesight, one of the issues with Noto I can show objectively: look at the "i"s. The dot is almost indistinguishable. And that's how the whole font feels to me.

**Comment by Calinou** [comment]
@samdze I keep the current work-in-progress branch here: https://github.com/Calinou/godot/tree/improve-editor-theme

It needs further tweaks to solve some quirks and avoid reducing the screen real estate available.

> Now, looking at Noto Sans' glyphs in Google Fonts does not feel as bad as it does in the editor. I can think of two things: there is some issue with rendering scaled down fonts in Godot; or Medium does it for me, and Noto does not have a Medium style.

Can you take a screenshot of Noto Sans in your browser at the same size as the default Godot font? This is probably due to different font hinting settings, which you can configure in the Editor Settings (on **Editor > Interface**).

**Comment by pycbouh** [comment]
@Calinou 
> Can you take a screenshot of Noto Sans in your browser at the same size as the default Godot font?

[Sure](https://user-images.githubusercontent.com/11782833/76420320-fd1baf80-63b2-11ea-8bec-72cb634b0b50.png). Though I still find that glyphs themselves are easier to read at a glance in case of IBM Plex. Things like a small curve at the bottom of lowercase "l" help me, personally.

[Changing hinting to "Normal"](https://user-images.githubusercontent.com/11782833/76420802-be3a2980-63b3-11ea-89d6-7c0392cf63d1.png) setting reduces some blurriness in the editor, but letter spacing suffers in some cases. Which is why Medium works for me, I guess. Glyphs just are a bit thicker, which reduces blur from "Light" hinting, while still having a more even spacing.

**Comment by dalexeev** [comment]
I think we can decrease the font size and padding a little. But I'm against hiding the setting values just to save space in the width.

**Comment by Calinou** [comment]
> I think we can decrease the font size and padding a little. But I'm against hiding the setting values just to save space in the width.

See https://github.com/godotengine/godot-proposals/issues/2980.

**Comment by SteampunkWizardStudios** [comment]
If it's at all possible I think a variable spacing option would probably be the best implementation. I personally prefer the wider spacing though.

**Comment by Calinou** [comment]
> If it's at all possible I think a variable spacing option would probably be the best implementation. I personally prefer the wider spacing though.

This is already available in 4.3 and later with the **Interface > Theme > Spacing Preset** editor settings. You can choose between Normal, Compact and Spacious.

**Comment by vaguesyntax** [comment]
Hi. this issue or feature request is almost a year old. And now I wanna talk more about it. I think Godot should go to a small UI revision. For example : 

![Image](https://github.com/user-attachments/assets/49399e88-f438-445c-ae1f-1dc1e4b163e9)  in this screenshot. Godot is loosing so much space. I think we should move the x,y,z input places to right side of the "Position" label. And we can do this for almost all of the Vector3 and also for the Vector2. And another topic is that I specificly don' know the reason but Godot is not looking modern when we compare it to UE or Unity. I think we should discuss about the reasons for it. For me Unreal Engine looks way more clean and compact but readable at the same time. And the last topic is unneccesery places that should be  replaced. For example : 

![Image](https://github.com/user-attachments/assets/d6f35143-6a3d-4e96-ae2c-ed9f929f5ad6)  I am using Godot for over 2 years and I don't remember any time me using this place in the inspector. And I also don't know that can I remove this place with an add-on or not. And probably there could be other places that is not used that often so we can put those settings in to popup. Let me know what you think.

**Comment by MunWolf** [comment]
Just wanted to put in my 2 cents, most of my frustration with the spacing in godot comes from all the space given to the 'label' side of the inspector

<img width="1118" height="1024" alt="Image" src="https://github.com/user-attachments/assets/92127837-3e1d-43dc-9ac7-fdda25496044" />

an example with a custom inspector plugin using add_property_editor


<img width="1130" height="472" alt="Image" src="https://github.com/user-attachments/assets/4cd0e4b5-4d5d-4045-95ac-11ac58784e91" />

Not exactly sure if you can easily fix this and also keep the size uniform but I would much rather have differences in label sizes between sections then have all of them take half the available horizontal space.

Here is an example with the dead space removed

<img width="770" height="434" alt="Image" src="https://github.com/user-attachments/assets/3579ce42-0205-42ad-86c6-355e7b00f690" />
